File: Trie/C/trie.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define ALPHABET_SIZE 26 // Assuming only lowercase letters a-z

// Trie Node structure
struct TrieNode
{
    struct TrieNode *children[ALPHABET_SIZE];
    bool isEndOfWord; // True if the word ends at this node
};

// Function to create a new Trie node
struct TrieNode *createNode()
{
    struct TrieNode *newNode = (struct TrieNode *)malloc(sizeof(struct TrieNode));
    if (newNode)
    {
        newNode->isEndOfWord = false;
        for (int i = 0; i < ALPHABET_SIZE; i++)
            newNode->children[i] = NULL;
    }
    return newNode;
}

// Function to insert a word into the Trie
void insert(struct TrieNode *root, const char *word)
{
    struct TrieNode *current = root;
    while (*word)
    {
        int index = *word - 'a';
        if (!current->children[index])
            current->children[index] = createNode();
        current = current->children[index];
        word++;
    }
    current->isEndOfWord = true;
}

// Function to search for a word in the Trie
bool search(struct TrieNode *root, const char *word)
{
    struct TrieNode *current = root;
    while (*word)
    {
        int index = *word - 'a';
        if (!current->children[index])
            return false;
        current = current->children[index];
        word++;
    }
    return current->isEndOfWord;
}

// Helper function to check if a node has any children
bool hasChildren(struct TrieNode *node)
{
    for (int i = 0; i < ALPHABET_SIZE; i++)
        if (node->children[i])
            return true;
    return false;
}

// Function to remove a word from the Trie
bool removeWord(struct TrieNode *root, const char *word)
{
    if (!root)
        return false;

    if (*word)
    {
        int index = *word - 'a';
        if (root->children[index] && removeWord(root->children[index], word + 1))
        {
            free(root->children[index]);
            root->children[index] = NULL;

            // If the current node is not the end of another word and has no children, delete it
            return (!root->isEndOfWord && !hasChildren(root));
        }
    }
    else if (root->isEndOfWord)
    {
        root->isEndOfWord = false;

        // If the node has no children, return true to delete it
        return !hasChildren(root);
    }
    return false;
}

// Function to free the entire Trie
void freeTrie(struct TrieNode *root)
{
    if (!root)
        return;
    for (int i = 0; i < ALPHABET_SIZE; i++)
        freeTrie(root->children[i]);
    free(root);
}

// Main function to test Trie operations
int main()
{
    struct TrieNode *root = createNode();

    // Insert words
    insert(root, "apple");
    insert(root, "banana");
    insert(root, "bat");
    insert(root, "ball");

    // Search words
    printf("Search apple: %s\n", search(root, "apple") ? "Found" : "Not Found");
    printf("Search bat: %s\n", search(root, "bat") ? "Found" : "Not Found");
    printf("Search ball: %s\n", search(root, "ball") ? "Found" : "Not Found");
    printf("Search cat: %s\n", search(root, "cat") ? "Found" : "Not Found");

    // Remove a word
    printf("\nRemoving 'bat'\n");
    removeWord(root, "bat");
    printf("Search bat: %s\n", search(root, "bat") ? "Found" : "Not Found");

    // Free the Trie
    freeTrie(root);
    return 0;
}

File: Trie/Java/Trie.java
class TrieNode {
    TrieNode[] children;
    boolean isEndOfWord;

    // Constructor
    public TrieNode() {
        children = new TrieNode[26]; // 26 lowercase English letters
        isEndOfWord = false;
    }
}

public class Trie {
    private TrieNode root;

    // Constructor: Initialize the Trie
    public Trie() {
        root = new TrieNode();
    }

    // Insert a word into the Trie
    public void insert(String word) {
        TrieNode current = root;
        for (char ch : word.toCharArray()) {
            int index = ch - 'a';
            if (current.children[index] == null) {
                current.children[index] = new TrieNode();
            }
            current = current.children[index];
        }
        current.isEndOfWord = true;
    }

    // Search for a word in the Trie
    public boolean search(String word) {
        TrieNode current = root;
        for (char ch : word.toCharArray()) {
            int index = ch - 'a';
            if (current.children[index] == null) {
                return false;
            }
            current = current.children[index];
        }
        return current.isEndOfWord;
    }

    // Helper function to check if a node has children
    private boolean hasChildren(TrieNode node) {
        for (TrieNode child : node.children) {
            if (child != null) return true;
        }
        return false;
    }

    // Remove a word from the Trie
    public boolean remove(String word) {
        return removeHelper(root, word, 0);
    }

    private boolean removeHelper(TrieNode current, String word, int depth) {
        if (current == null) return false;

        if (depth == word.length()) {
            if (!current.isEndOfWord) return false; // Word doesn't exist
            current.isEndOfWord = false;
            return !hasChildren(current); // Delete node if no children
        }

        int index = word.charAt(depth) - 'a';
        if (removeHelper(current.children[index], word, depth + 1)) {
            current.children[index] = null;
            return !current.isEndOfWord && !hasChildren(current);
        }
        return false;
    }

    // Main method to test Trie operations
    public static void main(String[] args) {
        Trie trie = new Trie();

        // Insert words
        trie.insert("apple");
        trie.insert("banana");
        trie.insert("bat");
        trie.insert("ball");

        // Search words
        System.out.println("Search apple: " + (trie.search("apple") ? "Found" : "Not Found"));
        System.out.println("Search bat: " + (trie.search("bat") ? "Found" : "Not Found"));
        System.out.println("Search ball: " + (trie.search("ball") ? "Found" : "Not Found"));
        System.out.println("Search cat: " + (trie.search("cat") ? "Found" : "Not Found"));

        // Remove a word
        System.out.println("\nRemoving 'bat'");
        trie.remove("bat");
        System.out.println("Search bat: " + (trie.search("bat") ? "Found" : "Not Found"));
    }
}
